src/hazardous/secret_stream/xchacha20poly1305.rs

```diff
@@ -0,0 +1,455 @@
+			"SecretStreamXChaCha20Poly1305  {{ key: [***OMITTED***], nonce: [***OMITTED***]",
+		)
+	}
+}
+
+impl SecretStreamXChaCha20Poly1305 {
+	fn get_key(&self) -> chacha20Key {
+		chacha20Key::from(self.key)
+	}
+	fn get_nonce(&self) -> chacha20Nonce {
+		chacha20Nonce::from(self.nonce)
+	}
+	fn get_inonce_mut(&mut self) -> &mut [u8] {
+		&mut self.nonce[SECRETSTREAM_XCHACHA20POLY1305_COUNTERBYTES..]
+	}
+	fn get_counter_mut(&mut self) -> &mut [u8] {
+		&mut self.nonce[..SECRETSTREAM_XCHACHA20POLY1305_COUNTERBYTES]
+	}
+
+	/// creates a new internal state
+	pub fn new(key: chacha20Key, nonce: Nonce) -> Self {
+		let mut s = SecretStreamXChaCha20Poly1305 {
+			key: hchacha20(&key, &nonce.as_ref()[..HCHACHA_NONCESIZE]).unwrap(),
+			nonce: [0u8; IETF_CHACHA_NONCESIZE],
+		};
+		s.reset_counter();
+		s.get_inonce_mut()
+			.copy_from_slice(&nonce.as_ref()[HCHACHA_NONCESIZE..]);
+		s
+	}
+
+	/// prepare a new push or pull operation
+	pub fn init(&mut self, key: chacha20Key, nonce: Nonce) {
+		*self = SecretStreamXChaCha20Poly1305::new(key, nonce);
+	}
+
+	#[must_use = "SECURITY WARNING: Ignoring a Result can have real security implications."]
+	/// derives a new secret key used for encryption/decryption
+	pub fn rekey(&mut self) -> Result<(), UnknownCryptoError> {
+		let mut new_key_and_inonce =
+			[0u8; CHACHA_KEYSIZE + SECRETSTREAM_XCHACHA20POLY1305_INONCEBYTES]; //TODO
+
+		new_key_and_inonce[..CHACHA_KEYSIZE].copy_from_slice(&self.key[..CHACHA_KEYSIZE]);
+		new_key_and_inonce[CHACHA_KEYSIZE..]
+			.copy_from_slice(&self.nonce[SECRETSTREAM_XCHACHA20POLY1305_COUNTERBYTES..]);
+
+		chacha20_enc_in_place(
+			&self.get_key(),
+			&self.get_nonce(),
+			0,
+			&mut new_key_and_inonce,
+		)?;
+
+		self.key[..CHACHA_KEYSIZE].copy_from_slice(&new_key_and_inonce[..CHACHA_KEYSIZE]);
+		self.nonce[SECRETSTREAM_XCHACHA20POLY1305_COUNTERBYTES..]
+			.copy_from_slice(&new_key_and_inonce[CHACHA_KEYSIZE..]);
+
+		self.reset_counter();
+		Ok(())
+	}
+
+	#[must_use = "SECURITY WARNING: Ignoring a Result can have real security implications."]
+	/// TODO: Doc
+	/// //Ecryptes a message
+	pub fn push(
+		&mut self,
+		plaintext: &[u8],
+		ad: Option<&[u8]>,
+		dst_out: &mut [u8],
+		tag: Tag,
+	) -> Result<(), UnknownCryptoError> {
+		const BLOCKSIZE: usize = 64;
+		let mut block = [0u8; BLOCKSIZE];
+		let mut slen = [0u8; 8];
+		let pad = [0u8; 16];
+		let adlen = match ad {
+			Some(v) => v.len(),
+			None => 0,
+		};
+		let msglen = plaintext.len();
+		let cipherpos = core::mem::size_of::<Tag>();
+		let macpos = cipherpos + msglen;
+		if dst_out.len() < SECRETSTREAM_XCHACHA20POLY1305_ABYTES + msglen {
+			return Err(UnknownCryptoError);
+		}
+
+		chacha20_enc_in_place(&self.get_key(), &self.get_nonce(), 0, &mut block)?;
+		let mut poly = init(&OneTimeKey::from_slice(&block[..POLY1305_KEYSIZE])?);
+		use zeroize::Zeroize;
+		block.zeroize();
+		if adlen > 0 {
+			poly.update(ad.unwrap())?;
+		}
+		poly.update(&pad[..((16usize.overflowing_sub(adlen).0) & 15)])?;
```

Comment:
There is already a padding function available in `chacha20poly1305`. If we make that public (`pub(crate)`), then we could use it like the other AEAD implementations do.

Reply:
It's now available on master.